### PR TITLE
Looking at match01 --- flakey test

### PR DESCRIPTION
**Pull request by KevinRansom**
Okay … what was happening is … the StableNameGenerator is global.  The fsharpqa tests use a compiler server to reduce build time.  Since the fsharpqa test framework parallelizes the tests and they can run in different orders, causing the stablenamegenerator to create different names, the 

This change attaches the stablenamegenerator to TcGlobals and plumb in tcGlobals, to a number of low level ast generation apis.

Because tcGlobals is recreated for every compilation, the Match01 optimization test case uses a bsl file, and so we noticed the name change.

I made a new type CompilerGlobalState to hang this type of state, because tcGlobals needs to appear after tast.fs.

Fixes https://github.com/microsoft/visualfsharp/issues/6657

### CODE REVIEW

File: tests/fsharpqa/Source/Optimizations/Inlining/env.lst [hunk 1]
```diff
@@ -1,4 +1,4 @@
 # flaky test: https://github.com/Microsoft/visualfsharp/issues/6657
```
Comment (brettfo) [inline tests/fsharpqa/Source/Optimizations/Inlining/env.lst]: Remove comment and add `fixes #6657` in notes

File: src/fsharp/xlf/FSComp.txt.es.xlf [hunk 2]
```diff
@@ -109,7 +109,7 @@
       </trans-unit>
       <trans-unit id="followingPatternMatchClauseHasWrongType">
         <source>All branches of a pattern match expression must return values of the same type as the first branch, which here is '{0}'. This branch returns a value of type '{1}'.</source>
-        <target state="translated">Todas las ramas de una expresión de coincidencia de patrón deben devolver valores del mismo tipo. La primera rama devolvió un valor de tipo "{0}", pero esta rama devolvió un valor de tipo "\{1 \}".</target>
+        <target state="new">All branches of a pattern match expression must return values of the same type as the first branch, which here is '{0}'. This branch returns a value of type '{1}'.</target>
```
Comment (dsyme) [inline src/fsharp/xlf/FSComp.txt.es.xlf]: odd change here

File: src/fsharp/tast.fs [hunk 3]
```diff
@@ -2543,7 +2535,9 @@ and
 and ValData = Val
 and [<NoEquality; NoComparison; StructuredFormatDisplay("{DebugText}")>]
     Val = 
-    { 
+    { /// Mutable for global state
+      mutable val_global_state: CompilerGlobalState option
```
Comment (dsyme) [inline src/fsharp/tast.fs]: This is a problem - it is adding a word to the critical Val data structure, which we have worked very hard to remove every word from.

I will search around for what makes this neccessary

File: src/fsharp/tast.fs [hunk 4]
```diff
@@ -2907,10 +2901,10 @@ and [<NoEquality; NoComparison; StructuredFormatDisplay("{DebugText}")>]
         //         let dt = System.DateTime.Now - System.DateTime.Now // IsMemberOrModuleBinding = false, IsCompiledAsTopLevel = true, IsMember = false, CompilerGenerated=true
         //    
         // However we don't need this for CompilerGenerated members such as the implementations of IComparable
-        if x.IsCompiledAsTopLevel && not x.IsMember && (x.IsCompilerGenerated || not x.IsMemberOrModuleBinding) then 
-            globalStableNameGenerator.GetUniqueCompilerGeneratedName(givenName, x.Range, x.Stamp) 
-        else 
-            givenName
+        match x.val_global_state with
```
Comment (dsyme) [inline src/fsharp/tast.fs]: I guess we should just pass `g` into `CompiledName`

**Comment by dsyme** [comment]
`val_global_state` only gets consumed here: https://github.com/microsoft/visualfsharp/pull/6673/files#diff-5a2b2c121409423e80d58b7ffaccd472R2904.  So I think just try passing `g` to `CompiledName` - it's probably always available as per the rest of the PR

File: src/fsharp/xlf/FSComp.txt.es.xlf (hunk 2 above)
Comment (KevinRansom) [reply src/fsharp/xlf/FSComp.txt.es.xlf]: Yes.

File: src/fsharp/tast.fs (hunk 4 above)
Comment (KevinRansom) [reply src/fsharp/tast.fs]: I would love to have passed around TcGlobals, however, it uses types from tast.fs, and so I can't pass tcGlobals that far back

**Comment by KevinRansom** [comment]
@dsyme, I will give that a look.

File: src/fsharp/tast.fs [hunk 5]
```diff
@@ -4561,7 +4552,7 @@ and
     [<DebuggerBrowsable(DebuggerBrowsableState.Never)>]
     member x.DebugText = x.ToString()
 
-    override x.ToString() = sprintf "TBind(%s, ...)" x.Var.CompiledName
+    override x.ToString() =  sprintf "TBind(%s, ...)" (x.Var.CompiledName None)
```
Comment (KevinRansom) [inline src/fsharp/tast.fs]: @ dsyme … This is why CompilerGlobalSources is an option.
There is no way to get tcglobals to here,

**Comment by KevinRansom** [comment]
@dsyme redone without the extra field.  I think there may be even more plumbing in this one, sigh !!!! but it doesn't bloat the val record.

Let me know what you think

**Comment by KevinRansom** [comment]
passed

**Comment by KevinRansom** [comment]
passed

**Comment by KevinRansom** [comment]
passed

**Comment by KevinRansom** [comment]
passed

**Comment by KevinRansom** [comment]
passed

**Comment by cartermp** [comment]
@KevinRansom It's passed for quite some time now...

**Comment by KevinRansom** [comment]
@cartermp, I know, but it was intermittent failing, I wanted to be sure … the fix really fixed it.

File: src/fsharp/tast.fs (hunk 5 above)
Comment (brettfo) [reply src/fsharp/tast.fs]: Nit: whitespace before `sprintf`.

**Pull request merged**